packages/fxa-auth-server/scripts/audit-orphaned-customers.ts

```diff
@@ -0,0 +1,267 @@
+  }
+  return false;
+};
+
+const parseDateForFirestore = (date: number) => Math.floor(date / 1000);
+
+const parseStartDate = (date: string): number =>
+  parseDateForFirestore(
+    date ? Date.parse(date) : Date.parse('2022-12-23T00:00:00')
+  );
+
+const parseEndDate = (date: string): number =>
+  parseDateForFirestore(
+    date ? Date.parse(date) : Date.parse('2022-12-31T23:59:59')
+  );
+
+const parseSubscriptionStatus = (
+  status: string
+): Stripe.Subscription['status'] | undefined => {
+  if (status) {
+    if (!isOfStatus(status)) {
+      throw new Error(
+        `Please provide a valid status. Invalid status: ${status}`
+      );
+    }
+    return status as Stripe.Subscription['status'];
+  }
+  return undefined;
+};
+
+function convertToCsv(data: CustomerData[]) {
+  const replacer = (key: string, value: any) => (value === null ? '' : value); // specify how you want to handle null values here
+  const header = Object.keys(data[0]);
+  return [
+    header.join(','), // header row first
+    ...data.map((row: CustomerData) =>
+      header
+        .map((fieldName: string) =>
+          JSON.stringify(row[fieldName as keyof typeof row], replacer)
+        )
+        .join(',')
+    ),
+  ].join('\r\n');
+}
+
+async function writeCsv(data: CustomerData[], file: string) {
+  if (!data.length) {
+    return;
+  }
+  const csv = convertToCsv(data);
+  try {
+    await fs.writeFile(`${parseDateForFirestore(Date.now())}_${file}.csv`, csv);
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+const errorsOutput: CustomerData[] = [];
+
+/**
+ * Get Stripe Customers with active subscriptions within a time period
+ * If the customer has a uid, add to StripeCustomers object
+ * If the customer does not have a uid, add to ErrorsOutput
+*/
+async function retrieveStripeCustomers(
+  startDate: number,
+  endDate: number,
+  status?: string,
+): Promise<Customers> {
+  const log = Container.get(AuthLogger);
+
+  log.info('retrieveStripeCustomers', { message: 'Checking Stripe customers...' });
+  const config = Container.get(AppConfig);
+  const stripe = new Stripe(config.subscriptions.stripeApiKey, {
+    apiVersion: '2022-08-01',
+    maxNetworkRetries: 3,
+  });
+  const baseQuery = `created<=${endDate} AND created>=${startDate}`;
+  const params: Stripe.SubscriptionSearchParams = {
+    query: status
+      ? `${baseQuery} AND status:'${status}'`
+      : `${baseQuery} AND -status:'canceled'`,
+    expand: ['data.customer', 'data.latest_invoice'],
+    limit: 100,
+  };
+
+  const stripeCustomers: Customers = {};
+
+  for await (const sub of stripe.subscriptions.search(params)) {
+    const uid =
+      isCustomer(sub.customer) && sub.customer.metadata['userid']
+        ? sub.customer.metadata['userid']
+        : undefined;
+    const email =
+      isCustomer(sub.customer) && sub.customer.email
+        ? sub.customer.email
+        : undefined;
+    const output: CustomerData = {
+      customer: sub.customer,
```

Comment:
Instead of the entire Customer, could we only return the ID? Makes reading the CSV easier.